### PR TITLE
Add support for BIT data type in MaterializedMySQL

### LINKED ISSUE #15182: MaterializeMySQL: support BIT data type
2020.09.23 16:13:20.829140 [ 24677 ] {} <Error> MaterializeMySQLSyncThread: Code: 1002, e.displayText() = DB::Exception: ParseMetaData: Unhandled data type:16, Stack trace (when copying this message, always include the lines below):

0. Poco::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) @ 0x114ffed0 in /usr/bin/clickhouse
1. DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) @ 0x8367afd in /usr/bin/clickhouse
2. ? @ 0xe2b4199 in /usr/bin/clickhouse
3. DB::MySQLReplication::TableMapEvent::parseImpl(DB::ReadBuffer&) @ 0xe2ae6b5 in /usr/bin/clickhouse
4. DB::MySQLReplication::MySQLFlavor::readPayloadImpl(DB::ReadBuffer&) @ 0xe2b2931 in /usr/bin/clickhouse
5. DB::MySQLProtocol::IMySQLReadPacket::readPayload(DB::ReadBuffer&, unsigned char&) @ 0xe2bdc81 in /usr/bin/clickhouse
6. DB::MySQLProtocol::PacketEndpoint::tryReceivePacket(DB::MySQLProtocol::IMySQLReadPacket&, unsigned long) @ 0xe2cc057 in /usr/bin/clickhouse
7. DB::MySQLClient::readOneBinlogEvent(unsigned long) @ 0xe2c9a81 in /usr/bin/clickhouse
8. DB::MaterializeMySQLSyncThread::synchronization(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) @ 0xe2a3b11 in /usr/bin/clickhouse
9. ? @ 0xe2a3f36 in /usr/bin/clickhouse
10. ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) @ 0x8395fcf in /usr/bin/clickhouse
11. ? @ 0x83948f3 in /usr/bin/clickhouse
12. start_thread @ 0x7ea5 in /usr/lib64/libpthread-2.17.so
13. __clone @ 0xfe8dd in /usr/lib64/libc-2.17.so
 (version 20.10.1.4709 (official build))

### PR DESCRIPTION
**Pull request by zzsmdfj**
Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added support for `BIT` data type in `MaterializedMySQL`. Closes #15182,  #32233

### CODE REVIEW

**Comment by stigsb** [comment]
I'm not able to click and write comments in the code, so I'll write it here.

On this line in materialize_with_ddl.py: 

```py
    assert "test_database_datatype" in clickhouse_node.query("SHOW DATABASES")
```

This may result in a flakey test, please do this instead:

```py
    check_query(clickhouse_node, "SHOW DATABASES LIKE 'test_database_datatype'", "test_database_datatype\n")
```

**Comment by zzsmdfj** [comment]
> I'm not able to click and write comments in the code, so I'll write it here.
> 
> On this line in materialize_with_ddl.py:
> 
> ```python
>     assert "test_database_datatype" in clickhouse_node.query("SHOW DATABASES")
> ```
> 
> This may result in a flakey test, please use `check_query`:
> 
> ```python
>     check_query(clickhouse_node, "SELECT name FROM system.tables WHERE database = 'test_database_datatype'", "t1\n")
> ```

ok, i will fix it. thanks mentioned that.

File: src/Core/MySQL/MySQLReplication.cpp [hunk 1]
```diff
@@ -584,6 +585,19 @@ namespace MySQLReplication
                         }
                         break;
                     }
+                    case MYSQL_TYPE_BIT:
+                    {
+                        UInt32 bits = ((meta >> 8) * 8) + (meta & 0xff);
```
Comment (tavplubix) [inline src/Core/MySQL/MySQLReplication.cpp:590]: Could you please explain what does it mean? What's the difference from `field_len`? Can it be greater that 64?

File: src/Core/MySQL/MySQLReplication.cpp [hunk 2]
```diff
@@ -584,6 +585,19 @@ namespace MySQLReplication
...
+                        UInt32 size = (bits + 7) / 8;
+
+                        Bitmap bitmap_value;
+                        String byte_buffer;
+                        byte_buffer.resize(size);
+                        readBigEndianStrict(payload, reinterpret_cast<char *>(byte_buffer.data()), size);
+                        readBitmapFromStr(byte_buffer.c_str(), bitmap_value, size);
+                        row.push_back(Field{UInt64{bitmap_value.to_ulong()}});
```
Comment (tavplubix) [inline src/Core/MySQL/MySQLReplication.cpp:598]: It it equivalent to simply casting `char *` to `UInt64 *`? I mean something like that:
```
UInt64 bitmap;
assert(size <= sizeof(bitmap));
readBigEndianStrict(payload, reinterpret_cast<char *>(&bitmap), size);
row.push_back(bitmap);
```

**Comment by tavplubix** [comment]
Integration tests flaky check has passed
Other integration tests failures are not related

File: src/Core/MySQL/MySQLReplication.cpp (hunk 1 above)
Comment (zzsmdfj) [reply src/Core/MySQL/MySQLReplication.cpp:590]: https://github.com/mysql/mysql-server/blob/3290a66c89eb1625a7058e0ef732432b6952b435/sql/log_event.cc#L1968 reference by this. 

The first byte of metadata is an integer divided by 8, and the second byte of metadata is the remainder of 8. size is  equivalently field_len（the value of field_len is not calculated above）

“The BIT data type is used to store bit values. A type of BIT(M) enables storage of M-bit values. M can range from 1 to 64.” reference from: https://dev.mysql.com/doc/refman/8.0/en/bit-type.html

File: src/Core/MySQL/MySQLReplication.cpp (hunk 2 above)
Comment (zzsmdfj) [reply src/Core/MySQL/MySQLReplication.cpp:598]: yes, the two ways of this are the same, however your as shown above  it looks more elegantly and concise. i will improve it.

Comment (zzsmdfj) [reply src/Core/MySQL/MySQLReplication.cpp]: done!

**Review by tavplubix — APPROVED** [review]

**Pull request merged**